File: tests/fixtures/typescript_project/calculator.ts
/**
 * Calculator module with basic arithmetic operations
 */

import { Point, CalculatorOptions } from './types/models';

/**
 * Adds two numbers together
 * @param a - First number
 * @param b - Second number
 * @returns The sum
 */
export function add(a: number, b: number): number {
    return a + b;
}

/**
 * Subtracts b from a
 * @param a - The minuend
 * @param b - The subtrahend
 * @returns The difference
 */
export function subtract(a: number, b: number): number {
    return a - b;
}

/**
 * Multiplies two numbers
 * @param a - First number
 * @param b - Second number
 * @returns The product
 */
export function multiply(a: number, b: number): number {
    return a * b;
}

/**
 * Divides a by b, returns null if b is zero
 * @param a - The dividend
 * @param b - The divisor
 * @returns The quotient or null
 */
export function divide(a: number, b: number): number | null {
    if (b === 0) {
        return null;
    }
    return a / b;
}

/**
 * Private helper function
 */
function _privateHelper(x: number): boolean {
    return x >= 0;
}

/**
 * Type for operation functions
 */
type Operation = (a: number, b: number) => number;

/**
 * Applies a custom operation to two numbers
 * @param a - First number
 * @param b - Second number
 * @param operation - Operation function
 * @returns Result of the operation
 */
export function applyOperation(a: number, b: number, operation: Operation): number {
    const result = operation(a, b);
    
    // Nested arrow function
    const formatter = (x: number): string => `Result is: ${x}`;
    
    console.log(formatter(result));
    return result;
}

/**
 * A complex operation with nested functions
 * @param base - The base value
 * @returns The result
 */
export function complexOperation(base: number): number {
    const multiplier = 2;
    
    // First nested function
    const double = (x: number): number => x * multiplier;
    
    // Nested function inside the first
    const applyTwice = (x: number): number => {
        const firstPass = double(x);
        
        // Another nested function
        const addBase = (y: number): number => y + base;
        
        return addBase(firstPass);
    };
    
    return applyTwice(base);
}

/**
 * Calculates the distance between two points
 * @param p1 - First point
 * @param p2 - Second point
 * @returns The distance
 */
export function pointDistance(p1: Point, p2: Point): number {
    const dx = p1.x - p2.x;
    const dy = p1.y - p2.y;
    return Math.sqrt(dx * dx + dy * dy);
}

/**
 * A simple calculator class
 * 
 * This class maintains state for calculator operations.
 */
export class Calculator {
    private _history: string[];
    public value: number;
    
    /**
     * Creates a new Calculator
     * @param initialValue - Starting value (default: 0)
     */
    constructor(initialValue: number = 0) {
        this.value = initialValue;
        this._history = [];
    }
    
    /**
     * Adds a number to the current value
     * @param n - Number to add
     * @returns The new value
     */
    add(n: number): number {
        this.value += n;
        this._history.push(`add ${n}`);
        return this.value;
    }
    
    /**
     * Subtracts a number from the current value
     * @param n - Number to subtract
     * @returns The new value
     */
    subtract(n: number): number {
        this.value -= n;
        this._history.push(`subtract ${n}`);
        return this.value;
    }
    
    /**
     * Resets the calculator to zero
     */
    reset(): void {
        this.value = 0;
        this._history = [];
    }
    
    /**
     * Gets the operation history
     * @returns Copy of the history
     */
    getHistory(): string[] {
        return [...this._history];
    }
    
    /**
     * Private helper method
     */
    private _logOperation(op: string): void {
        this._history.push(op);
    }
    
    toString(): string {
        return `Calculator(value: ${this.value})`;
    }
}

File: tests/fixtures/rust_project/src/calculator.rs
//! Calculator module with basic arithmetic operations

use crate::models::{Calculator, Point};
use std::ops::Add;

/// Adds two numbers together
///
/// # Examples
/// ```
/// let result = add(2, 3);
/// assert_eq!(result, 5);
/// ```
pub fn add(a: i32, b: i32) -> i32 {
    a + b
}

/// Subtracts b from a
///
/// # Arguments
/// * `a` - The minuend
/// * `b` - The subtrahend
///
/// # Returns
/// The difference a - b
pub fn subtract(a: i32, b: i32) -> i32 {
    a - b
}

/// Multiplies two numbers
pub fn multiply(a: i32, b: i32) -> i32 {
    a * b
}

/// Divides a by b, returns None if b is zero
pub fn divide(a: i32, b: i32) -> Option<i32> {
    if b == 0 {
        None
    } else {
        Some(a / b)
    }
}

/// A function that uses a closure
pub fn apply_operation<F>(a: i32, b: i32, op: F) -> i32
where
    F: Fn(i32, i32) -> i32,
{
    let formatter = |x: i32| -> String { format!("Result is: {}", x) };
    let result = op(a, b);
    println!("{}", formatter(result));
    result
}

/// Uses the Calculator struct from models
pub fn create_calculator() -> Calculator {
    Calculator::new()
}

/// Creates a calculator with an initial value
pub fn create_calculator_with_value(value: i32) -> Calculator {
    Calculator::with_value(value)
}

/// Performs a sequence of operations on a calculator
pub fn perform_sequence(mut calc: Calculator, operations: &[(i32, char)]) -> i32 {
    for (value, op) in operations {
        match op {
            '+' => {
                calc.add(*value);
            }
            '-' => {
                calc.subtract(*value);
            }
            _ => {}
        }
    }
    calc.value
}

/// Calculates the distance between two points
pub fn point_distance(p1: Point, p2: Point) -> f64 {
    let dx = (p1.x - p2.x) as f64;
    let dy = (p1.y - p2.y) as f64;
    (dx * dx + dy * dy).sqrt()
}

/// A complex operation with nested closures
pub fn complex_operation(base: i32) -> i32 {
    let multiplier = 2;

    // First closure
    let double = |x: i32| -> i32 { x * multiplier };

    // Nested closure inside the first
    let apply_twice = |x: i32| -> i32 {
        let first_pass = double(x);
        // Another nested closure
        let add_base = |y: i32| -> i32 { y + base };
        add_base(first_pass)
    };

    apply_twice(base)
}
